File: src/utils/fuelCalculations.ts
import type { FuelLog } from '../types/database';

/**
 * Calculate average mileage from full tank logs
 * Formula: (LogB.odometer - LogA.odometer) / LogB.litres
 * where LogA and LogB are the last two consecutive full tank logs
 */
export function calculateMileage(fuelLogs: FuelLog[]): number | null {
  // Filter to only full tank logs, sorted by odometer descending
  const fullTankLogs = fuelLogs
    .filter((log) => log.is_full_tank)
    .sort((a, b) => b.odometer - a.odometer);

  if (fullTankLogs.length < 2) {
    // Need at least 2 full tank logs to calculate mileage
    return null;
  }

  // Get the last two consecutive full tank logs
  const logB = fullTankLogs[0]; // Most recent
  const logA = fullTankLogs[1]; // Previous

  if (!logA || !logB || logB.litres === 0) {
    return null;
  }

  // Calculate mileage: (distance between refuels) / (litres in second refuel)
  const mileage = (logB.odometer - logA.odometer) / logB.litres;

  // Return null if result is invalid (negative or unrealistic)
  if (mileage <= 0 || mileage > 1000) {
    // 1000 km/L is unrealistic, likely a data error
    return null;
  }

  return Math.round(mileage * 100) / 100; // Round to 2 decimal places
}

/**
 * Get the most recent fuel price from fuel logs
 */
export function getLastFuelPrice(fuelLogs: FuelLog[]): number | null {
  if (fuelLogs.length === 0) {
    return null;
  }

  // Sort by date descending, then by created_at descending
  const sortedLogs = [...fuelLogs].sort((a, b) => {
    const dateCompare = new Date(b.date).getTime() - new Date(a.date).getTime();
    if (dateCompare !== 0) return dateCompare;
    return new Date(b.created_at).getTime() - new Date(a.created_at).getTime();
  });

  return sortedLogs[0]?.price_per_litre ?? null;
}

File: lib/core/utils/fuel_calculations.dart
// Fuel calculation utilities — ported from React's fuelCalculations.ts.
// These must match the React implementation exactly.

/// A minimal fuel log representation for calculations.
class FuelLogEntry {
  const FuelLogEntry({
    required this.odometer,
    required this.litres,
    required this.pricePerLitre,
    required this.isFullTank,
    required this.date,
    required this.createdAt,
  });

  final double odometer;
  final double litres;
  final double pricePerLitre;
  final bool isFullTank;
  final String date; // YYYY-MM-DD
  final DateTime createdAt;
}

/// Calculate average mileage from the last 10 full-tank fill-ups.
///
/// Uses consecutive full-tank pairs: km/l = (odoB - odoA) / litresB.
/// Averages across all valid pairs within the last 10 full-tank logs.
double? calculateMileage(List<FuelLogEntry> fuelLogs) {
  // Filter to only full tank logs, sorted by odometer descending, take last 10
  final fullTankLogs = fuelLogs.where((log) => log.isFullTank).toList()
    ..sort((a, b) => b.odometer.compareTo(a.odometer));

  final recent = fullTankLogs.take(10).toList();

  if (recent.length < 2) {
    return null;
  }

  double totalMileage = 0;
  int validPairs = 0;

  for (int i = 0; i < recent.length - 1; i++) {
    final logB = recent[i];
    final logA = recent[i + 1];

    if (logB.litres == 0) continue;

    final mileage = (logB.odometer - logA.odometer) / logB.litres;

    // Skip invalid pairs (negative or unrealistic)
    if (mileage <= 0 || mileage > 1000) continue;

    totalMileage += mileage;
    validPairs++;
  }

  if (validPairs == 0) return null;

  // Round to 2 decimal places
  return ((totalMileage / validPairs) * 100).roundToDouble() / 100;
}

/// Get the most recent fuel price from fuel logs.
///
/// Sorts by date descending, then by createdAt descending.
double? getLastFuelPrice(List<FuelLogEntry> fuelLogs) {
  if (fuelLogs.isEmpty) {
    return null;
  }

  final sortedLogs = List<FuelLogEntry>.from(fuelLogs)
    ..sort((a, b) {
      final dateCompare = b.date.compareTo(a.date);
      if (dateCompare != 0) return dateCompare;
      return b.createdAt.compareTo(a.createdAt);
    });

  return sortedLogs.first.pricePerLitre;
}
